### PR TITLE
Added `reuse-executions` property to pipeline schema

### PR DESCRIPTION
**Pull request by dangquangdon**

### CODE REVIEW

File: valohai_yaml/schema/pipeline.yaml
Comment (tarekoraby) [inline valohai_yaml/schema/pipeline.yaml:41]: Would it make sense to call it `node-reuse` with true/false values?

Comment (tarekoraby) [reply valohai_yaml/schema/pipeline.yaml:41]: Or `step-reuse`

**Review by akx — DISMISSED** [review]
1. Tests fail.
2. The .gitignore changes seem unrelated?
3. What are the semantics here – is this just the default for the pipeline API `allow_reuse` flag (which, afaiu, only affects reusing executions into execution nodes?), or something else? I can't seem to find a ref to any sort of spec for this.

**Comment by dangquangdon** [comment]
> 3\. What are the semantics here – is this just the default for the pipeline API `allow_reuse` flag (which, afaiu, only affects reusing executions into execution nodes?), or something else? I can't seem to find a ref to any sort of spec for this.

The idea is that if user specify this in `valohai.yaml`, the Automatically reuse switch button in the UI will also reflect it

File: valohai_yaml/schema/pipeline.yaml
Comment (dangquangdon) [reply valohai_yaml/schema/pipeline.yaml:41]: we called it `allow_reuse` in roi's code too so we could keep it for consistency

Comment (tarekoraby) [reply valohai_yaml/schema/pipeline.yaml:41]: > we called it allow_reuse in roi's code 

Do you mean in the internal code or in the user-facing UI?

What we call it internally doesn't matter to the end users. In the UI we had some preamble text explaining what the feature does and how it works. In `valohai.yaml`, the terminology we use should be self-explanatory, and `allow-reuse` isn't descriptive enough of what the feature does.

File: valohai_yaml/schema/pipeline.yaml
Comment (tarekoraby) [inline valohai_yaml/schema/pipeline.yaml:47]: Somewhere in the description, we should mention the conditions that lead to execution being matched (i.e., that the execution is in the same project, that it has the same inputs, parameters, code commit, and data).

**Comment by hylje** [comment]
The snapshots have changed due to a change in a downstream library version.

I've fixed the snapshot issue in https://github.com/valohai/valohai-yaml/pull/147 which is waiting for review.

File: valohai_yaml/schema/pipeline.yaml
Comment (dangquangdon) [reply valohai_yaml/schema/pipeline.yaml:41]: sure, let's use `node-reuse`/`step-reuse` as you suggested. However, IMO, from user's perspective, it doens't matter that much because whichever keyword we use, it wouldn't be self-explanatory enough for this feature, user will have to look for the description and/or documentation anyway to be able to understand what the keyword does.

File: valohai_yaml/schema/pipeline.yaml
Comment (dangquangdon) [reply valohai_yaml/schema/pipeline.yaml:47]: Could you suggest some text for it? Tbh I'm not good with the wording 😂

**Comment by dangquangdon** [comment]
> The snapshots have changed due to a change in a downstream library version that changed sort order.
> 
> I've fixed the snapshot issue in #147 which is waiting for review.

Nice, thanks for the info. Did it fail also in your local machine when you were fixing it? Because it didn't happen in mine 🤔

**Comment by hylje** [comment]
It didn't fail at first but when I upgraded the dependencies it started failing the same way it did in CI (which always installs freshest dependencies)

**Comment by dangquangdon** [comment]
> It didn't fail at first but when I upgraded the dependencies it started failing the same way it did in CI (which always installs freshest dependencies)

I checked the dependencies versions as well and they were the same as in CI, but after your comment, I remove the whole `venv` and create a fresh new one and was able to reproduce it. Seemed like it was some kind of caching issue for me.

File: valohai_yaml/schema/pipeline.yaml
Comment (dangquangdon) [reply valohai_yaml/schema/pipeline.yaml]: added some more text to it 👍

File: valohai_yaml/schema/pipeline.yaml
Comment (akx) [reply valohai_yaml/schema/pipeline.yaml]: TBH, I liked `allow-reuse` more. `node-reuse` is inaccurate, since it does not reuse nodes, but executions that would be executed by some nodes. `reuse-executions` smells like it would be most accurate for what we have now.

**Review by akx — COMMENTED** [review]
LGTM (not approve-gating yet), if @tarekoraby is OK with the new name.

**Comment by tarekoraby** [comment]
`reuse-executions` LGTM 👍

**Review by akx — APPROVED** [review]

**Pull request merged**
